### PR TITLE
Feature/add associated clients tlv to topology response

### LINKED ISSUE #704: [TASK] Add Associated Clients TLV to Topology Response message
## Description

Add Associated Clients TLV to Topology Response message.

## Exit criteria

Topology Response message contains Associated Clients TLV with real data.

## Detailed description

For a detailed description, see task #525 (this task has been divided into smaller tasks to be managed more easily).

### PR DESCRIPTION
**Pull request by mariomaz**
[TASK] Add Associated Clients TLV to Topology Response message #704

### CODE REVIEW

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:62]: The commit message mentions that ruid is removed as well. However, that was already done in an earlier commit, so the commit message should be updated.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1953]: (nitpick) for clarity it would be better to call this radio_mac.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1957]: Note that radio_uid has to stay radio_uid since that's the name used by the MultiAP Spec.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: Excellent documentation, and excellent choice of std::array as the underlying type. :1st_place_medal:

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Why this intermediate array rather than writing immediately to `m_radio_info_map[request->iface_mac()].supported_channels` ?

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Can't we use std::copy?
```
    std::copy(channels, beerocks::message::SUPPORTED_CHANNELS_LENGTH,
              m_radio_info_map[request->iface_mac()].supported_channels);
```

File: agent/src/beerocks/slave/tlvf_utils.cpp
Comment (arnout) [inline agent/src/beerocks/slave/tlvf_utils.cpp:9]: Nitpick: I think this would fit better in bcl than here.

File: agent/src/beerocks/slave/tlvf_utils.cpp
Comment (arnout) [inline agent/src/beerocks/slave/tlvf_utils.cpp:19]: Again, radio_mac would be better than ruid (nitpick).

File: agent/src/beerocks/slave/tlvf_utils.cpp
Comment (arnout) [inline agent/src/beerocks/slave/tlvf_utils.cpp]: IIRC somebody made a change that makes it possible to allocate a list of size 0, so this condition would no longer be needed.

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/son_slave_thread.cpp:3435]: It would be better to change hostap_params.supported_channels into an array (or vector) as well, but that can be done later.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: Since the MAC address here has to be one of the VAPs in vaps_list, it would make sense to do something similar to the radioInfo, i.e. make a struct that contains bssid, ssid and associated clients. Note that it wouldn't be a map then, because we need to be able to index it both based on BSSID and on SSID. So a vector and searching with find() is more appropriate.

However, that's probably a bit too big of a change, so let's leave that for some later refactoring.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Why the `&&` reference? A normal reference seems more appropriate to me. But I'm not so familiar with the semantics of `&&` references (I don't even know what they're called :-))

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I'm not sure that this is the correct thing to do. AFAIU it is possible for a client to be associated in multiple BSSIDs. And in any case, when it disassociates, a disconnect message should be received from it.

If it is necessary to be able to find the one BSSID with which a station is associated, we should probably find another solution for that.

That said: probably better keep it as is now and fix it if there really is a problem.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Doesn't this check both sides of the tuple? How can the timestamp ever be identical?

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I think the timestamp should be marked as early as possible, i.e. it should be part of the message at least. Preferably it would already be set in bwl - but that would be a bigger change so not necessary here.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Since the time_point is definitely different from the one that was inserted, this will never find anything.

I think we need a helper function find_associated_client to find the client. Or it has to be a map client_mac -> time_point.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Again, why `&&`? `const auto &radio_info_entry` seems more appropriate.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Same here.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2025]: For completeness, we should also handle the `elapsed < 0` case.

**Review by arnout — APPROVED** [review]
I have a *lot* of comments, mostly nitpicks but a few important ones.

Please fix at least the important ones, rebase on master, wait for CI to finish and merge right away. This really needs to be tested close to master so don't wait too long after the rebase before merging.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml
Comment (mariomaz) [reply common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_backhaul.yaml:62]: Done

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1953]: Done

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1957]: Ok

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: Thanks

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Done

File: agent/src/beerocks/slave/tlvf_utils.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/tlvf_utils.cpp:9]: Let's see how this file evolves before moving it. Since it is a nitpick comment, I'm not going to move it for the moment.

File: agent/src/beerocks/slave/tlvf_utils.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/tlvf_utils.cpp:19]: This parameter is copied to a TLV field named radio_uid, so should not be renamed.
Description is "Radio unique identifier of the radio for which capabilities are reported."

File: agent/src/beerocks/slave/tlvf_utils.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/tlvf_utils.cpp]: Removed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: Ok. Not changed now

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Removed

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/son_slave_thread.cpp:3435]: Ok, later

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: This is a C++11 rvalue reference. 
AFAIK, auto&& is the same as const auto&.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: There is a list of associated clients per BSSID. BSSID value is being received as part of the message and it is used to obtain the list of associated clients to that BSSID. So AFAIK there shouldn't be any problem.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: associated_clients is an std::unordered_set so std::find works on the hash value of the key. 
associated_client_hash has been defined so the hash value is computed as the hash value of the MAC address of the associated client. The timestamp is not even being considered to compare the tuples.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Ok, not changed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: AFAIK they are equivalent and && is preferred over const auto&. 
I'm sorry but don't know how to explain it. That's something I learned once and use whenever I have to iterate over a map. I can change it if you prefer.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Sorry, I was miserably wrong. And it was over complicated so I've also simplified it

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: You are right! Fixed with a map client_mac -> time_point

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2025]: Done

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Thanks for the info.

I read all of [this article](http://thbecker.net/articles/rvalue_references/section_01.html) to understand what rvalue references are about. As far as I can see, it makes little sense to declare a variable as an rvalue reference. There's a [stackoverflow answer](https://stackoverflow.com/questions/18766324/is-it-useless-to-declare-a-local-variable-as-rvalue-reference-e-g-t-r-move) that says the same but I trust it a little less. 

As far as I can see, `const auto&` and `auto&&` are indeed more or less equivalent. However, the `const auto&` pattern is used all over the code, and is (IMO) easier to understand for people who are not utter C++11 experts. Also our CONTRIBUTING.md contains an example that uses that pattern. So I tend to prefer the `const auto&` variant even if they are equivalent.

**Comment by arnout** [comment]
The `auto &&` thing is a nitpick so I removed the don't merge tag. We still need a second review however.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I think you missed my point.

When you receive a message that a client has associated with a BSS, you remove that client from any previous BSS it was associated with. My point is that:
- maybe it actually still is associated with that BSS;
- in any case, when it does disassociate, we'll get a disassociation message and it will be removed then.

In other words, it is not needed and maybe incorrect.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: This can be removed now (nitpick)

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1844]: Shouldn't we add an error message here?

File: framework/tlvf/tlvf_conf.yaml
Comment (tomereli) [inline framework/tlvf/tlvf_conf.yaml:31]: The commit message should mention which TLVs are added (nitpick)

File: controller/src/beerocks/master/son_master_thread.cpp
Comment (tomereli) [inline controller/src/beerocks/master/son_master_thread.cpp:1338]: We should also print the capabilities here. Then simulate the capabilities in the agent dummy implementation, and then make sure that test_flows.sh chech that the capability report was generated. But that can be in a different PR, so (nitpick)

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:245]: It's really great that you write such a comprehensive description of the new container.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I agree with @arnout - so please change this to `const auto &`.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I really prefer that you change it to `const auto &`. Pretty please?:)

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: same here

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: same here

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: and here

**Review by tomereli — APPROVED** [review]
Approving already cause my comments are pretty simple can can be fixed.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Done

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:1844]: Error messages are already being written to log inside the add_ap_radio_basic_capabilities method, so there is no need to do it here back again.

File: framework/tlvf/tlvf_conf.yaml
Comment (mariomaz) [reply framework/tlvf/tlvf_conf.yaml:31]: Done

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:245]: Thanks

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Done

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Done

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Done

File: controller/src/beerocks/master/son_master_thread.cpp
Comment (mariomaz) [reply controller/src/beerocks/master/son_master_thread.cpp:1338]: I agree with you. All tests in test_flows.sh should be extended to test the contents of the messages and not only if a particular message name appears on the log. Tests should also check if the message has been sent or received and the number of times.

**Pull request merged**
